src/core/IncomingRequestManager.cpp

```diff
@@ -246,29 +436,49 @@ void IncomingContactRequest::setConnection(ContactRequestServer *c)
     connection = c;
     emit hasActiveConnectionChanged();
 }
+#endif
 
 void IncomingContactRequest::accept(ContactUser *user)
 {
     qDebug() << "Accepting contact request from" << m_hostname;
 
-    /* Create the contact */
-    if (!user)
-    {
+    // Create the contact if necessary
+    if (!user) {
         Q_ASSERT(!nickname().isEmpty());
         user = manager->contacts->addContact(nickname());
         user->setHostname(QString::fromLatin1(m_hostname));
     }
 
+#ifdef PROTOCOL_NEW
+    using namespace Protocol::Data::ContactRequest;
+
+    // If we have a connection, send the response and pass it to ContactUser
+    if (connection) {
+        auto channel = connection->findChannel<Protocol::ContactRequestChannel>();
+        if (channel) {
+            // Channel will close after sending a final response
+            user->assignConnection(connection.data());
+            channel->setResponseStatus(Response::Accepted);
+
+            if (connection->parent() != user) {
+                BUG() << "ContactUser didn't claim connection from incoming contact request";
+                connection->close();
+            }
+        } else {
+            connection->close();
+        }
+        connection.clear();
+    }
+#else
     user->settings()->write("remoteSecret", Base64Encode(remoteSecret()));
 
-    /* If there is a connection, send the accept message and morph it to a primary connection */
-    if (connection)
-    {
+    if (connection) {
         connection.data()->sendAccept(user);
         connection = (ContactRequestServer*)0;
```

Comment:
If you're using C++11 (which you seem to be doing), then you could start using nullptr in some places where it makes sense.